src/storage/src/hummock/compactor/compactor_runner.rs

```diff
@@ -227,6 +236,58 @@ impl CompactorRunner {
     }
 }
 
+pub fn partition_overlapping_sstable_infos(
+    mut origin_infos: Vec<SstableInfo>,
+) -> Vec<Vec<SstableInfo>> {
+    pub struct SstableGroup {
+        ssts: Vec<SstableInfo>,
+        max_right_bound: Vec<u8>,
+    }
+
+    impl PartialEq for SstableGroup {
+        fn eq(&self, other: &SstableGroup) -> bool {
+            self.max_right_bound == other.max_right_bound
+        }
+    }
+    impl PartialOrd for SstableGroup {
+        fn partial_cmp(&self, other: &SstableGroup) -> Option<std::cmp::Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+    impl Eq for SstableGroup {}
+    impl Ord for SstableGroup {
+        fn cmp(&self, other: &SstableGroup) -> std::cmp::Ordering {
+            // Pick group with the smallest right bound for every new sstable.
+            KeyComparator::compare_encoded_full_key(&other.max_right_bound, &self.max_right_bound)
+        }
+    }
+    let mut groups: BinaryHeap<SstableGroup> = BinaryHeap::default();
+    origin_infos.sort_by(|a, b| {
+        let x = a.key_range.as_ref().unwrap();
+        let y = b.key_range.as_ref().unwrap();
+        KeyComparator::compare_encoded_full_key(&x.left, &y.left)
+    });
+    for sst in origin_infos {
+        // Pick group with the smallest right bound for every new sstable. So do not check the larger one if the smallest one does not meet condition.
+        if let Some(mut prev_group) = groups.peek_mut() {
+            if KeyComparator::encoded_full_key_less_than(
+                &prev_group.max_right_bound,
+                &sst.key_range.as_ref().unwrap().left,
+            ) {
+                prev_group.max_right_bound = sst.key_range.as_ref().unwrap().right.clone();
+                prev_group.ssts.push(sst);
+                continue;
+            }
+        }
+        groups.push(SstableGroup {
+            max_right_bound: sst.key_range.as_ref().unwrap().right.clone(),
+            ssts: vec![sst],
+        });
+    }
+    assert!(!groups.is_empty());
+    groups.into_iter().map(|group| group.ssts).collect_vec()
```

Comment:
IMO, we can add a warning log to determine the efficiency of the grouping algorithm. (If sst is all unordered, the grouping algorithm may fail, so that we can find problems faster).

question:  What can we do if the grouping algorithm fails?

Reply:
I add a log info after calling this method